### DISCUSSION TITLE
writeHead(301) in getInitialProps doesn't work with Link components

### DISCUSSION THREAD

**Discussion by marshallku** [vercel/next.js#47349] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000
Binaries:
  Node: 18.12.1
  npm: 9.6.2
  Yarn: 1.22.19
  pnpm: N/A
Relevant packages:
  next: 12.2.2
  eslint-config-next: 12.2.2
  react: 17.0.2
  react-dom: 17.0.2
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true), Data fetching (gS(S)P, getInitialProps)

### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/pedantic-cartwright-9uoxiq?file=%2Fpages%2F_document.tsx&selection=%5B%7B%22endColumn%22%3A9%2C%22endLineNumber%22%3A4%2C%22startColumn%22%3A9%2C%22startLineNumber%22%3A4%7D%5D

### To Reproduce

1. Add `res.writeHead(302, URL_TO_REDIRECT)` in `getInitialProps` of page component
2. Access that link with `Link` component

### Describe the Bug

Redirect never works unless you access that page directly.

### Expected Behavior

Redirected to `URL_TO_REDIRECT`

**Comment by balazsorban44** [comment]
You forgot to end the response with `res.end()`, `writeHead` does not do that. It will just hang indefinitely.

You might want to check [Middleware](https://nextjs.org/docs/advanced-features/middleware) as a better alternative to `getInitialProps` maybe.

**Reply by marshallku** [reply]
Sorry @balazsorban44, I did miswrite code to reproduce my issue.
It never works even with `res.end()` too. ;(

**Comment by chaitanyaj2222** [comment]
Hey @marshallku , would you be able to tell me why you considered using `getInitialProps` here instead of `getServerSideProps`.  Are there any specific reasons behind it?

**Reply by marshallku** [reply]
Hi @7777chaitanya, I update my code in codesandbox to make it easier to understand the situation.
Please check it if you mind.

Thanks.

**Comment by icyJoseph** [comment]
Hi,

Isn't the problem that you are doing an SPA transition into a page with getInitialProps? Those run client side!

So there are no req/res objects. Instead you gotta redirect with the Router instance.

There's also an issue with not calling the pages's getInitialProps IIRC.

```tsx
import App, { AppProps, AppContext } from "next/app";
import Router from "next/router";

const MyApp = ({ Component, pageProps }: AppProps) => (
  <Component {...pageProps} />
);

MyApp.getInitialProps = async (context: AppContext) => {
  const ctx = await App.getInitialProps(context);

  const { res, query } = ctx;

  if (query?.redirect === "1") {
    if (res) {
      console.log("Redirect query exists, redirecting...");
      res?.writeHead(302, { Location: "/" })?.end();
    } else {
      Router.replace("/");
    }
  }

  return ctx;
};

export default MyApp;
```

And your GetInitialPropsPage should be something like:

```tsx
import { NextPageContext } from "next";
import Link from "next/link";
import Router from "next/router";

const GetInitialPropsPage = () => (
  <div>
    <h1>GetInitialProps</h1>
    <Link href="/">Back</Link>
  </div>
);

GetInitialPropsPage.getInitialProps = async (ctx: NextPageContext) => {
  const { res } = ctx;

  if (res) {
    res?.writeHead(302, { Location: "/" });
    res?.end();
  } else {
    Router.replace("/");
  }
  return ctx;
};

export default GetInitialPropsPage;

```
